quinn-proto/src/connection/assembler.rs

```diff
@@ -145,15 +181,25 @@ impl Assembler {
             return;
         }
 
-        self.data.push(Buffer { offset, bytes });
-        // Why 32: on the one hand, we want to defragment rarely, ideally never
+        let size = bytes.len();
+        self.buffered += size;
+        self.allocated += allocation_size;
+        self.data.push(Buffer {
+            offset,
+            bytes,
+            size,
+            allocation_size,
+        });
+        // Rationale: on the one hand, we want to defragment rarely, ideally never
         // in non-pathological scenarios. However, a pathological or malicious
         // peer could send us one-byte frames, and since we use reference-counted
         // buffers in order to prevent copying, this could result in keeping a lot
-        // of memory allocated. In the worst case scenario of 32 1-byte chunks,
-        // each one from a ~1000-byte datagram, using 32 limits us to having a
-        // maximum pathological over-allocation of about 32k bytes.
-        if self.data.len() - self.defragmented > 32 {
+        // of memory allocated. This limits over-allocation in proportion to the
+        // buffered data. The constants are chosen somewhat arbitrarily and try to
+        // balance between defragmentation overhead and over-allocation.
+        let over_allocation = (self.allocated - self.buffered) as f32;
+        let threshold = ((self.buffered as f32) * 1.5f32).max(4096f32);
```

Comment:
```suggestion
        let threshold = ((self.buffered as f32) * 1.5).max(4096.0);
```

It also seems like an overallocation of only 4096 bytes shouldn't be enough to trigger this (given the performance impact we have seen). Should go back up to something more like 32k?

Reply:
Yeah, makes sense. Fixed.